File: admin_list_controls/src/admin_list_controls/actions.js
import * as constants from "./constants";

export function create_action(action) {
    switch(action.action_type) {
        case 'toggle_panel':
            return {
                type: constants.TOGGLE_PANEL,
                ref: action.ref,
            };
        case 'collapse_panel':
            return {
                type: constants.COLLAPSE_PANEL,
                ref: action.ref,
            };
        case 'set_value':
            return {
                type: constants.SET_VALUE,
                name: action.name,
                value: action.value,
            }
        case 'remove_value':
            return {
                type: constants.REMOVE_VALUE,
                name: action.name,
                value: action.value,
            }
        case 'submit_form':
            return {
                type: constants.SUBMIT_FORM,
            };
        case 'clear_search_input':
            return {
                type: constants.CLEAR_SEARCH_INPUT,
            };
        case 'link':
            return {
                type: constants.LINK,
                url: action.url,
            };
        default:
            console.error('Unknown action type', action);
            throw new Error(`Unknown action ${action.action_type}`);
    }
}

File: admin_list_controls/actions.py
class BaseAction:
    object_type = 'action'
    action_type = ''

    def serialize(self):
        return {
            'object_type': self.object_type,
            'action_type': self.action_type,
        }


class SetValue(BaseAction):
    action_type = 'set_value'

    def __init__(self, name, value):
        self.name = name
        self.value = value

    def serialize(self):
        return dict(super().serialize(), **{
            'name': self.name,
            'value': self.value,
        })


class RemoveValue(BaseAction):
    action_type = 'remove_value'

    def __init__(self, name, value):
        self.name = name
        self.value = value

    def serialize(self):
        return dict(super().serialize(), **{
            'name': self.name,
            'value': self.value,
        })


class Link(BaseAction):
    action_type = 'link'

    def __init__(self, url):
        self.url = url

    def serialize(self):
        return dict(super().serialize(), **{
            'url': self.url,
        })


class TogglePanel(BaseAction):
    action_type = 'toggle_panel'
    show_panel_toggle_icon = True

    def __init__(self, ref):
        self.ref = ref

    def serialize(self):
        return dict(super().serialize(), **{
            'ref': self.ref,
            'show_panel_toggle_icon': self.show_panel_toggle_icon,
        })


class CollapsePanel(BaseAction):
    action_type = 'collapse_panel'

    def __init__(self, ref):
        self.ref = ref

    def serialize(self):
        return dict(super().serialize(), **{
            'ref': self.ref,
        })


class ClearSearchInput(BaseAction):
    action_type = 'clear_search_input'


class SubmitForm(BaseAction):
    action_type = 'submit_form'
